### ISSUE TITLE
Stale Object Modification Error

### ISSUE THREAD

**Issue by ichekrygin** [crossplaneio/crossplane#395]
* Bug Report
It appears we hitting concurrent update issue on a given object:
```
"error": "failed to update status of CRD instance postgresql-b00a1c60-5d60-11e9-9440-9cb6d08bde99: 
	Operation cannot be fulfilled on cloudsqlinstances.database.gcp.crossplane.io \"postgresql-b00a1c60-5d60-11e9-9440-9cb6d08bde99\": 
		the object has been modified; please apply your changes to the latest version and try again",
"stacktrace": 
	"github.com/crossplaneio/crossplane/vendor/github.com/go-logr/zapr.(*zapLogger).Error
	/home/illya/go/src/github.com/crossplaneio/crossplane/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/crossplaneio/crossplane/pkg/controller/gcp/database.(*Reconciler).Reconcile
	/home/illya/go/src/github.com/crossplaneio/crossplane/pkg/controller/gcp/database/cloudsql_instance.go:214\ngithub.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/home/illya/go/src/github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/home/illya/go/src/github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1
	/home/illya/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/home/illya/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait.Until
	/home/illya/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"
```

I stumbled over this issue few times on different resources. 
This could be an intermittent issue on my dev environment.

**Comment by HashedDan** [comment]
I have also encountered this a number of times.

**Comment by ichekrygin** [comment]
So far I observed this happening on `cloudsql_instag.go` controller (gcp/database). 
I suspect this is related to issuing multiple Updates on the object w/out re-fetching it from the API.
The object (database instance) becomes "stale" after the first Update:
https://github.com/crossplaneio/crossplane/blob/master/pkg/controller/gcp/database/cloudsql_instance.go#L192
thus the second update fails:
https://github.com/crossplaneio/crossplane/blob/master/pkg/controller/gcp/database/cloudsql_instance.go#L213

**Comment by negz** [comment]
Linking back to this related conversation - https://github.com/crossplaneio/gitlab-controller/pull/4#discussion_r283047018.

I believe @ichekrygin validated our theory that calling `Update()` will refresh the updated object with its latest resource version, so in theory multiple sequential `Update()` operations on the same object should work, as long as something else has not updated the object out of band.

**Comment by ichekrygin** [comment]
I thought I did. I ran another experiment.
I added the following function:
```go
func (r *Reconciler) Update(ctx context.Context, obj *databasev1alpha1.CloudsqlInstance) error {
	log.Info("***********************") // so it is easy to spot :)
	log.Info("Update before", "rv", obj.ObjectMeta.ResourceVersion, "gen", obj.ObjectMeta.Generation)
	err := r.Client.Update(ctx, obj)
	log.Info("Update after", "rv", obj.ObjectMeta.ResourceVersion, "gen", obj.ObjectMeta.Generation)
	return err
}
```
This function is called every time we call  `r.Update` and prints resource version and generation before and immediately after the update. 
The results are quite inconsistent. In some cases, the resource version is incremented while in others it isn't.

```
2019-05-14T13:07:48.810-0700	INFO	crossplane.controller.cloudsqlinstances.database.gcp.crossplane.io	***********************
2019-05-14T13:07:48.810-0700	INFO	crossplane.controller.cloudsqlinstances.database.gcp.crossplane.io	Update before	{"rv": "603205", "gen": 9}
2019-05-14T13:07:48.816-0700	DEBUG	kubebuilder.controller	Successfully Reconciled	{"controller": "cloudmemorystoreinstances.cache.gcp.crossplane.io", "request": "crossplane-system/redis-e39997eb-7681-11e9-bfa6-9cb6d08bde99"}
2019-05-14T13:07:48.816-0700	INFO	crossplane.controller.cloudsqlinstances.database.gcp.crossplane.io	Update after	{"rv": "603205", "gen": 9}
2019-05-14T13:07:48.816-0700	DEBUG	crossplane.controller.cloudmemorystoreinstances.cache.gcp.crossplane.io	reconciling	{"kind": "cloudmemorystoreinstance.cache.gcp.crossplane.io/v1alpha1", "request": "crossplane-system/redis-e39997eb-7681-11e9-bfa6-9cb6d08bde99"}
2019-05-14T13:07:48.818-0700	DEBUG	kubebuilder.controller	Successfully Reconciled	{"controller": "CloudsqlInstance-controller", "request": "crossplane-system/postgresql-e3995b62-7681-11e9-bfa6-9cb6d08bde99"}
2019-05-14T13:07:48.899-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	***********************
2019-05-14T13:07:48.899-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	Update before	{"rv": "603529", "gen": 33}
2019-05-14T13:07:48.903-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	Update after	{"rv": "603581", "gen": 34}
2019-05-14T13:07:48.903-0700	DEBUG	crossplane.controller.gke.compute.gcp.crossplane.io	reconciling	{"kind": "gkecluster.compute.gcp.crossplane.io/v1alpha1", "request": "crossplane-system/gke-e399276d-7681-11e9-bfa6-9cb6d08bde99"}
2019-05-14T13:07:49.014-0700	DEBUG	crossplane.controller.bucket.storage.crossplane.io	reconciling	{"kind": "bucket.storage.crossplane.io/v1alpha1", "request": "crossplane-system/demo-lfs"}
2019-05-14T13:07:49.122-0700	DEBUG	crossplane.controller.bucket.storage.gcp.crossplane.io	reconciling	{"kind": "bucket.storage.gcp.crossplane.io/v1alpha1", "request": "crossplane-system/gcs-e39c12df-7681-11e9-bfa6-9cb6d08bde99"}
2019-05-14T13:07:49.152-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	***********************
2019-05-14T13:07:49.152-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	Update before	{"rv": "603581", "gen": 34}
2019-05-14T13:07:49.158-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	Update after	{"rv": "603582", "gen": 35}
2019-05-14T13:07:49.158-0700	DEBUG	crossplane.controller.gke.compute.gcp.crossplane.io	reconciling	{"kind": "gkecluster.compute.gcp.crossplane.io/v1alpha1", "request": "crossplane-system/gke-e399276d-7681-11e9-bfa6-9cb6d08bde99"}
2019-05-14T13:07:49.223-0700	DEBUG	kubebuilder.controller	Successfully Reconciled	{"controller": "cloudmemorystoreinstances.cache.gcp.crossplane.io", "request": "crossplane-system/redis-e39997eb-7681-11e9-bfa6-9cb6d08bde99"}
2019-05-14T13:07:49.223-0700	DEBUG	crossplane.controller.cloudmemorystoreinstances.cache.gcp.crossplane.io	reconciling	{"kind": "cloudmemorystoreinstance.cache.gcp.crossplane.io/v1alpha1", "request": "crossplane-system/redis-e39997eb-7681-11e9-bfa6-9cb6d08bde99"}
2019-05-14T13:07:49.294-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	***********************
2019-05-14T13:07:49.294-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	Update before	{"rv": "603582", "gen": 35}
2019-05-14T13:07:49.300-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	Update after	{"rv": "603582", "gen": 35}
```
...
```
2019-05-14T13:11:49.347-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	***********************
2019-05-14T13:11:49.347-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	Update before	{"rv": "603732", "gen": 36}
2019-05-14T13:11:49.357-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	Update after	{"rv": "603882", "gen": 37}
2019-05-14T13:11:49.357-0700	DEBUG	crossplane.controller.gke.compute.gcp.crossplane.io	reconciling	{"kind": "gkecluster.compute.gcp.crossplane.io/v1alpha1", "request": "crossplane-system/gke-e399276d-7681-11e9-bfa6-9cb6d08bde99"}
2019-05-14T13:11:49.599-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	***********************
2019-05-14T13:11:49.599-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	Update before	{"rv": "603882", "gen": 37}
2019-05-14T13:11:49.602-0700	INFO	crossplane.controller.gke.compute.gcp.crossplane.io	Update after	{"rv": "603882", "gen": 37}
```

There appears to be no rhyme nor reason why this is happening.
I am going to follow up w/ `controller-runtime` to see if there is an open issue, and if not, will create one.

**Comment by ichekrygin** [comment]
So it appears `controller-runtime` can dish out "stale" reads :( , which is unusual for my case since I am running in local mode and have a single instance of `etcd` and `kubeapi`.
```
019-05-14T13:36:22.396-0700    DEBUG    gitlab.controller.gitlab-controller    reconciling resource claims
2019-05-14T13:36:24.186-0700    INFO    gitlab.controller.gitlab-controller    *************
2019-05-14T13:36:24.186-0700    INFO    gitlab.controller.gitlab-controller    Before    {"meta": "606382"}
2019-05-14T13:36:24.187-0700    INFO    gitlab.controller.gitlab-controller    After    {"meta": "606386"} <--- NEW 
2019-05-14T13:36:24.187-0700    INFO    gitlab.controller.gitlab-controller    one or more claims have pending status    {"reconcile action": "wait"}
2019-05-14T13:36:24.187-0700    DEBUG    gitlab.controller.gitlab-controller    reconciling    {"request": "crossplane-system/demo"}
2019-05-14T13:36:24.187-0700    DEBUG    gitlab.controller.gitlab-controller    reconciling resource claims
2019-05-14T13:36:25.990-0700    INFO    gitlab.controller.gitlab-controller    *************
2019-05-14T13:36:25.991-0700    INFO    gitlab.controller.gitlab-controller    Before    {"meta": "606382"} <--- STALE
2019-05-14T13:36:25.995-0700    INFO    gitlab.controller.gitlab-controller    After    {"meta": "606382"}
```

Makes me wonder if it is the case of `controller-runtime` caching issues (if any)?

**Comment by ichekrygin** [comment]
https://github.com/kubernetes-sigs/controller-runtime/issues/403

**Comment by jbw976** [comment]
related to #99

**Comment by jbw976** [comment]
@ichekrygin do you think your merged PR #556 completely takes care of this stale object modification issue? Or is there still more work that we should be tracking in this issue or a new issue?

**Comment by hasheddan** [comment]
I experienced this error outside of the GCP CloudsqlInstance context so I believe the issue will need to remain open.
